File: server/server_main/app/services/credit_service.py
from ..extensions import db
from ..models.user import User
from ..models.credit_transaction import CreditTransaction


def award_credits(user: User, amount: int, ref_type: str, ref_id: int) -> CreditTransaction:
    """Add credits to user balance and record the transaction."""
    user.credit_balance += amount
    txn = CreditTransaction(
        user_id=user.id,
        type="EARN",
        amount=amount,
        balance_after=user.credit_balance,
        ref_type=ref_type,
        ref_id=ref_id,
    )
    db.session.add(txn)
    return txn


def spend_credits(user: User, amount: int, ref_type: str, ref_id: int) -> CreditTransaction:
    """Deduct credits from user balance and record the transaction.

    Raises ValueError if balance is insufficient.
    """
    if user.credit_balance < amount:
        raise ValueError(
            f"Insufficient credits: balance={user.credit_balance}, required={amount}"
        )
    user.credit_balance -= amount
    txn = CreditTransaction(
        user_id=user.id,
        type="SPEND",
        amount=amount,
        balance_after=user.credit_balance,
        ref_type=ref_type,
        ref_id=ref_id,
    )
    db.session.add(txn)
    return txn

File: server/server_main/src/services/creditService.ts
import prisma from '../prisma'
import { getSettings, n } from './settingsService'

export async function creditsForVolume(volumeMl: number): Promise<number> {
  const s = await getSettings()
  if (volumeMl <= n(s, 'credit_tier_s_max_ml')) return n(s, 'credit_tier_s_credits')
  if (volumeMl <= n(s, 'credit_tier_m_max_ml')) return n(s, 'credit_tier_m_credits')
  return n(s, 'credit_tier_l_credits')
}

export async function awardCredits(
  userId: number, amount: number, refType: string, refId: number
) {
  const user = await prisma.user.update({
    where: { id: userId },
    data: { creditBalance: { increment: amount } },
  })
  return prisma.creditTransaction.create({
    data: { userId, type: 'EARN', amount, balanceAfter: user.creditBalance, refType, refId },
  })
}

export async function spendCredits(
  userId: number, amount: number, refType: string, refId: number
) {
  const user = await prisma.user.findUniqueOrThrow({ where: { id: userId } })
  if (user.creditBalance < amount) throw new Error('Insufficient credits')
  const updated = await prisma.user.update({
    where: { id: userId },
    data: { creditBalance: { decrement: amount } },
  })
  return prisma.creditTransaction.create({
    data: { userId, type: 'SPEND', amount, balanceAfter: updated.creditBalance, refType, refId },
  })
}
